File: Lintcode/79.Longest_common_substring.py
class Solution:
    # @param A, B: Two string.
    # @return: the length of the longest common substring.
    def longestCommonSubstring(self, A, B):
        # write your code here
        '''
        if len(A)==0 or len(B)==0:
            return 0
            
        sub_a = []
        for i in range(len(A)):
            if i == 0:
                for index in range(len(A)):
                    sub_a.append(A[index])
            else:
                for start in range(len(A)-i):
                    tempstring = ''
                    for j in range(start,start+i+1):
                        tempstring += A[j]
                    sub_a.append(tempstring)
                    
        for item in sub_a[::-1]:
            if item in B:
                return len(item)
        '''
        
        ans = 0
        
        for i in xrange(len(A)):
            for j in xrange(len(B)):
                l = 0
                while i + l < len(A) and j + l < len(B) and A[i + l] == B[j + l]:
                    l += 1
                if l > ans:
                    ans = l
        return ans
                

File: Lintcode/79.Longest_common_substring.cpp
class Solution {
public:    
    /**
     * @param A, B: Two string.
     * @return: the length of the longest common substring.
     */
    int longestCommonSubstring(string &A, string &B) {
        // write your code here
        /*
        int maxlen = 0;
        int xlen = A.size();
        int ylen = B.size();
        for (int i=0;i<xlen;i++)
        {
            for(int j=0;j<ylen;j++)
            {
                int temp = 0;
                while(A[i+temp]==B[j+temp] and (i+temp<xlen) and (j+temp<ylen))
                    temp += 1;
                if (temp > maxlen)
                    maxlen = temp;
            }
        }
        return maxlen;
        */
        
        if (A.empty() || B.empty())
            return 0;
        
        vector<string> sub_A;
        for (int i=0; i<A.size(); i++)
        {
            if (i>0)
            {
                for(int index=0;index<A.size()-i;index++)
                    sub_A.push_back(A.substr(index,i+1));
            }
            else
            {
                for(int index=0;index<A.size();index++)
                {
                    string temp = "";
                    temp += A[index];
                    sub_A.push_back(temp);
                }
            }
        }
        
        for (int index=sub_A.size()-1;index>=0;index--)
        {
            if(B.find(sub_A[index])<B.size())
                return sub_A[index].size();
        }
    }
};